### PR TITLE
AzureTestPlanTask: Implemented support for Jest

### PR DESCRIPTION
**Pull request by t-shreysingh**
**Task name**: AzureTestPlanTask: Implemented support for Jest

**Description**: Successfully installing jest-junit package and running test cases from different files.
https://dev.azure.com/t-shreysingh/Go%20project/_build/results?buildId=315&view=logs&j=275f1d19-1bd8-5591-b06b-07d489ea915a&t=ec886807-3279-5bc4-fce4-1ee4f1a980f1
Facing issues in publishing the junitfile in Test Tab. 

**Documentation changes required:** (Y/N) <Please mark if documentation changes are required>

**Added unit tests:** (Y/N) <Please mark if unit tests were added or updated according changes>

**Attached related issue:** (Y/N) <Please add link to related issue here>

**Checklist**:
- [ ] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [ ] Checked that applied changes work as expected
image
image

### CODE REVIEW

**Comment by t-shreysingh** [comment]
Close it

**Comment by t-shreysingh** [comment]
Reopen

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts [hunk 1]
```diff
@@ -1,14 +1,39 @@
...
+//Jest command like >set JEST_JUNIT_OUTPUT_NAME=TEST-Jest0-junit.xml
+//>npx jest --ci --reporters=default --reporters=jest-junit -t "JestTestName"
 export async function executeJestTests(testsToBeExecuted: string[]): Promise<number> {
 
-    // jest execution will be added
-    /*executable = npm
-    args = test
-    */
+    let finalStatus = 0; 
+    let jestNpm = tl.which("npm", true);
```
Comment (triptijain2112) [inline Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts:12]: why jestNpm?

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts [hunk 2]
```diff
@@ -1,14 +1,39 @@
...
+    await executeJestCommand(jestNpm, constants.INSTALL_JESTJUNIT);
```
Comment (triptijain2112) [inline Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts:13]: add comment why are we installing it

File: Tasks/AzureTestPlanV0/utils.ts [hunk 3]
```diff
@@ -42,4 +42,15 @@ export function separateGoTestName(inputString) {
         // If there is no dot in the string, return the original string
         return inputString;
     }
+}
+
+export function separateJestTestName(inputString) {
+    const lastDotIndex = inputString.lastIndexOf('.');
+
+    if (lastDotIndex !== -1) {
+        const stringWith = inputString.slice(lastDotIndex + 1);
```
Comment (triptijain2112) [inline Tasks/AzureTestPlanV0/utils.ts:51]: rename variable name stringWith

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts (hunk 1 above)
Comment (t-shreysingh) [reply Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts:12]: Just to know that this package will be used to install jest package.

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts [hunk 4]
```diff
@@ -1,14 +1,39 @@
...
 
-    console.log("jest changes1");
-    return 1;
+    //testsToBeExecuted: DescribeD TestH.DescribeD TestH,DescribeD TestG.DescribeD TestG,DescribeC TestF.DescribeC TestF,DescribeC TestE.DescribeC TestE,DescribeB TestD.DescribeB TestD,DescribeB TestC.DescribeB TestC,DescribeA TestB.DescribeA TestB,DescribeA TestA.DescribeA TestA
```
Comment (triptijain2112) [inline Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts:15]: trim this or add better sample for testsToBeExecuted

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts [hunk 5]
```diff
@@ -1,14 +1,39 @@
...
+    let jestNpx = tl.which("npx", true);
+    let i = 0;
+    for (let tests of testsToBeExecuted) {
+        const JestTestName = utils.separateJestTestName(tests);
+        try {
+            let junitFileName: string = `TEST-Jest${i}-junit.xml`;
+            tl.setVariable('JEST_JUNIT_OUTPUT_NAME', junitFileName);
+            let junitName = tl.getVariable('JEST_JUNIT_OUTPUT_NAME');
+            tl.debug(`Junit Filename ${junitName} executed successfully.`);
```
Comment (triptijain2112) [inline Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts:24]: Junit Filename ${junitName} environment set successfully.

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts [hunk 6]
```diff
@@ -1,14 +1,39 @@
...
+          
+            const jestCommand = `jest --ci --reporters=default --reporters=jest-junit -t "${JestTestName}"`;
+            const status = await executeJestCommand(jestNpx, jestCommand);
+            if (status != 0) {
+                finalStatus = 1;
+            }
+            tl.debug(`Test case ${JestTestName} executed successfully.`);
```
Comment (triptijain2112) [inline Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts:31]: we can move this log to console.log

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts (hunk 1 above)
Comment (t-shreysingh) [reply Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts:12]: Changed to npmPath.

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts (hunk 4 above)
Comment (t-shreysingh) [reply Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts:15]: github link

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts (hunk 6 above)
Comment (t-shreysingh) [reply Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts:31]: Added

File: Tasks/AzureTestPlanV0/utils.ts (hunk 3 above)
Comment (t-shreysingh) [reply Tasks/AzureTestPlanV0/utils.ts]: done

Comment (t-shreysingh) [reply Tasks/AzureTestPlanV0/utils.ts]: Resolved

Comment (t-shreysingh) [reply Tasks/AzureTestPlanV0/utils.ts]: Resolved

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts [hunk 7]
```diff
@@ -1,14 +1,39 @@
...
+    let i = 0;
+    for (let tests of testsToBeExecuted) {
+        const JestTestName = utils.separateJestTestName(tests);
+        try {
+            let junitFileName: string = `TEST-Jest${i}-junit.xml`;
+            tl.setVariable('JEST_JUNIT_OUTPUT_NAME', junitFileName);
+            let junitName = tl.getVariable('JEST_JUNIT_OUTPUT_NAME');
+            tl.debug(`Junit Filename ${junitName} executed successfully.`);
+          
+            const jestCommand = `jest --ci --reporters=default --reporters=jest-junit -t "${JestTestName}"`;
```
[the remark this answers is not in the document]
Comment (t-shreysingh) [reply Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts:26]: done

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts (hunk 2 above)
Comment (t-shreysingh) [reply Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts]: Done

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts (hunk 5 above)
Comment (t-shreysingh) [reply Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts]: Added

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts [hunk 8]
```diff
@@ -1,14 +1,39 @@
...
+    await executeJestCommand(npmPath, constants.INSTALL_JESTJUNIT);
+    //testToBeExecuted: <TestSuiteName1> <TestCase1>. <TestSuiteName1> <TestCase1>,<TestSuiteName2> <TestCase3>. <TestSuiteName2> <TestCase3>
+    let npxPath = tl.which("npx", true);
+    let i = 0;
+    for (let tests of testsToBeExecuted) {
+        const JestTestName = utils.separateJestTestName(tests);
+        try {
+            let junitFileName: string = `TEST-Jest${i}-junit.xml`;
+            tl.setVariable('JEST_JUNIT_OUTPUT_NAME', junitFileName);
+            let junitName = tl.getVariable('JEST_JUNIT_OUTPUT_NAME');
```
Comment (triptijain2112) [inline Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts:23]: just a thought: should we add a check if junitName is same as junitFileName?

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts (hunk 6 above)
Comment (triptijain2112) [reply Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts:31]: i don't see change made

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts (hunk 8 above)
Comment (t-shreysingh) [reply Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts:23]: They both are same just the variable names are different.

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts [hunk 9]
```diff
@@ -1,14 +1,39 @@
...
+            console.log(`Junit Filename ${junitName} environment set successfully.`);
```
Comment (adityashahms) [inline Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts:24]: What is the intention for this console log?

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts [hunk 10]
```diff
@@ -1,14 +1,39 @@
...
-    /*executable = npm
-    args = test
-    */
-
-    console.log("jest changes1");
-    return 1;
+    //Jest-Junit Link:https://github.com/jest-community/jest-junit
+    let finalStatus = 0; 
+    let npmPath = tl.which("npm", true);
+    await executeJestCommand(npmPath, constants.INSTALL_JESTJUNIT);
```
Comment (adityashahms) [inline Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts:14]: Please handle the installation failure using try catch block and bail out if faced any?

Also, please add debug logs to enable faster investigation as well

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts (hunk 9 above)
Comment (t-shreysingh) [reply Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts]: Changed it to tl.debug

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts [hunk 11]
```diff
@@ -1,14 +1,39 @@
...
+            console.log(`Junit Filename ${junitName} environment set successfully.`);
+          
+            const jestCommand = `jest --ci --reporters=default --reporters=jest-junit -t "${JestTestName}"`;
+            const status = await executeJestCommand(npxPath, jestCommand);
+            if (status != 0) {
+                finalStatus = 1;
+            }
+            tl.debug(`Test case ${JestTestName} executed successfully.`);
+        } catch (error) {
+            tl.debug(`Error executing ${JestTestName} test case: ${error}`);
```
[the remark this answers is not in the document]
Comment (t-shreysingh) [reply Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts]: Added

File: Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts (hunk 10 above)
Comment (t-shreysingh) [reply Tasks/AzureTestPlanV0/Invokers/jestinvoker.ts]: Added error handling for installation and debug logs

**Review by triptijain2112 — APPROVED** [review]

**Pull request merged**